File: typescript/resistor-color-trio/resistor-color-trio.ts
export const colorCode = (color:Color): number =>
  COLORS.indexOf(color);

type Color = typeof COLORS[number]

export const COLORS =
  [ 'black', 'brown', 'red', 'orange', 'yellow', 'green', 'blue', 'violet', 'grey', 'white' ] as const

export const decodedResistorValue = (colors: Color[]): string => {
  const number = (10 * colorCode(colors[0]) + colorCode(colors[1]))
  const zeros = colorCode(colors[2])

  switch(zeros) {
    case 2: case 3: case 4:
      return `${number * (10 ** (zeros-3))} kiloohms`
    case 5: case 6: case 7:
      return `${number * (10 ** (zeros-6))} megaohms`
    case 8: case 9:
      return `${number * (10 ** (zeros-9))} gigaohms`
    default:
      return `${number * (10 ** zeros)} ohms`;
  };
}

File: fsharp/resistor-color-trio/ResistorColorTrio.fs
module ResistorColorTrio

let toNumber = function
  | "black" -> 0
  | "brown" -> 1
  | "red" -> 2
  | "orange" -> 3
  | "yellow" -> 4
  | "green" -> 5
  | "blue" -> 6
  | "violet" -> 7
  | "grey" -> 8
  | "white" -> 9
  | _ -> failwith "unknown color"

let label (colors: string list) =
    let root = 10 * (toNumber colors[0]) + (toNumber colors[1])
    let zeros = pown 10 (toNumber colors[2])
    let number = uint64 root * uint64 zeros

    if number >= 1_000_000_000UL then $"{number / 1_000_000_000UL } gigaohms"
    elif number >= 1_000_000UL then $"{number / 1_000_000UL } megaohms"
    elif number >= 1_000UL then $"{number / 1_000UL } kiloohms"
    else $"{number} ohms"
